juju/juju
# Revoke leases

## Description of change

The highlight is that when a unit transition to dead, its leadership lease is revoked.
There's 4 main bits of work:

1. Update the leadership worker - we only use a raft based lease store now which auto expires leases so remove the manual expire api. Add revoke api.
2. Update the raft fsm to support revoking leases.
3. Remote legacy lease code in state and consolidate remaining legacy code needed foe upgrades.
(move code from updates.go to leaseupdates.go)
4. Update the api facades to remove the unit leadership when the unit goes to dead.

## QA steps

Deploy 3 units
watch juju status
remove the leader unit
see that juju status shows the transition to a new leader as soon as the deleted leader is gone

## Bug reference

https://bugs.launchpad.net/bugs/1469731
